### PR TITLE
Using dws-test-driver for DWS state progression integration tests

### PR DESCRIPTION
**Pull request by nathandotleeathpe**
Signed-off-by: Nathan Lee <nathan.lee@hpe.com>

### CODE REVIEW

File: src/burst_buffer/burst_buffer.lua [hunk 1]
```diff
@@ -163,7 +163,7 @@ function DWS:initialize(wlmID, jobID, userID, groupID, dw_directives, labels)
 	else
 		dwd_block = "dwDirectives:\n"
 		for k, v in ipairs(dw_directives) do
-			dwd_block = dwd_block .. "  - " .. dw_directives[k] .. "\n"
+			dwd_block = dwd_block .. "  - \"" .. dw_directives[k] .. "\"\n"
```
Comment (roehrich-hpe) [inline src/burst_buffer/burst_buffer.lua:166]: Sorry about that.

**Comment by roehrich-hpe** [comment]
"nathandotleeathpe force-pushed the main  branch from 793d98e to c5a29c8 Compare 4 hours ago"

When you do the force-push, I can't use github gui to look at what you've changed since the last time I was here.

**Comment by roehrich-hpe** [comment]
Update .gitignore for these pycache dirs:
```
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	testsuite/integration/src/tests/__pycache__/
	testsuite/integration/src/tests/dws_bb_plugin/__pycache__/
	testsuite/integration/src/tests/environment/__pycache__/

```

**Comment by roehrich-hpe** [comment]
I'm finished reviewing and testing this, up to commit c5a29c8.

**Comment by roehrich-hpe** [comment]
> Update .gitignore for these pycache dirs:
> 
> ```
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
> 	testsuite/integration/src/tests/__pycache__/
> 	testsuite/integration/src/tests/dws_bb_plugin/__pycache__/
> 	testsuite/integration/src/tests/environment/__pycache__/
> ```


And "reports/" as well:
```
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	testsuite/integration/reports/
	testsuite/integration/src/tests/__pycache__/
	testsuite/integration/src/tests/dws_bb_plugin/__pycache__/
	testsuite/integration/src/tests/environment/__pycache__/
```

File: testsuite/integration/src/features/test_dws_states.feature [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            #DW <state> action=wait
+            #DW Teardown action=wait
+            /bin/hostname
+
+        When the job is run
+        And a Workflow is created for the job
+        And the Workflow and job progress to the <state> state
+        And the job is canceled with the hurry flag set to <hurry_flag>
+        Then the Workflow and job progress to the Teardown state
+        And the Workflow's hurry flag is set to <hurry_flag>
```
[the remark this answers is not in the document]
Comment (nathandotleeathpe) [reply testsuite/integration/src/features/test_dws_states.feature:100]: As written, this test doesn't really work. I will probably refactor it when I work on the implementation next week. I'll take another look at your suggestion at that point.

File: testsuite/integration/src/tests/slurmctld.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    @retry(
+        wait=wait_fixed(6),
+        stop=stop_after_attempt(10),
+        reraise=True
+    )
+    def submit_job(self,scriptPath):
+        print("submit job")
+        # The --wait option could be used here. However, other tests need to
+        # asynchronously track the job status
+        cmd = "sbatch --output=%(s)s.out --error=%(s)s.error.out %(s)s " % {"s": scriptPath}
```
Comment (roehrich-hpe) [inline testsuite/integration/src/tests/slurmctld.py:55]: This is old-style string formatting.  We should be using the `.format()` method or f-strings.

**Comment by roehrich-hpe** [comment]
The docker-compose from the 'setup-kind' target...

```
Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
[+] Running 2/3
 ⠿ Network slurm_default  Created                                          0.0s
 ⠿ Container slurmctld    Started                                          0.5s
 ⠿ Container c1           Waiting                                        195.0s
```

The containers ....

```
$ docker container ls -a | grep slurm
a9d3d975faed   slurm-bb:test          "/usr/local/bin/dock…"   About a minute ago   Exited (1) About a minute ago                               c1
58213079cb9f   slurm-bb:test          "/usr/local/bin/dock…"   About a minute ago   Exited (1) About a minute ago                               slurmctld

$ docker logs slurmctld
---> Starting the MUNGE Authentication service (munged) ...
munged: Error: Failed to check keyfile "/etc/munge/munge.key": Permission denied

$ docker logs c1       
---> Starting the MUNGE Authentication service (munged) ...
munged: Error: Failed to check keyfile "/etc/munge/munge.key": Permission denied
```

**Comment by roehrich-hpe** [comment]
> The docker-compose from the 'setup-kind' target...
> 
> ```
> Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
> [+] Running 2/3
>  ⠿ Network slurm_default  Created                                          0.0s
>  ⠿ Container slurmctld    Started                                          0.5s
>  ⠿ Container c1           Waiting                                        195.0s
> ```
> 
> The containers ....
> 
> ```
> $ docker container ls -a | grep slurm
> a9d3d975faed   slurm-bb:test          "/usr/local/bin/dock…"   About a minute ago   Exited (1) About a minute ago                               c1
> 58213079cb9f   slurm-bb:test          "/usr/local/bin/dock…"   About a minute ago   Exited (1) About a minute ago                               slurmctld
> 
> $ docker logs slurmctld
> ---> Starting the MUNGE Authentication service (munged) ...
> munged: Error: Failed to check keyfile "/etc/munge/munge.key": Permission denied
> 
> $ docker logs c1       
> ---> Starting the MUNGE Authentication service (munged) ...
> munged: Error: Failed to check keyfile "/etc/munge/munge.key": Permission denied
> ```

Add a `docker volume rm slurm_etc_munge slurm_var_log_slurm` to the 'clean' target in `testsuite/integration/Makefile`.  That fixed my probs.

**Review by roehrich-hpe — APPROVED** [review]
Add the 'docker volume rm' to the 'clean' target.  I have no other issues.

**Pull request merged**